src/core/ddsi/src/q_entity.c

```diff
@@ -311,12 +311,22 @@ nn_vendorid_t get_entity_vendorid (const struct entity_common *e)
   return NN_VENDORID_UNKNOWN;
 }
 
-void ddsi_make_writer_info(struct ddsi_writer_info *wrinfo, const struct entity_common *e, const struct dds_qos *xqos)
+void ddsi_make_writer_info(struct ddsi_writer_info *wrinfo, const struct entity_common *e, const struct dds_qos *xqos, nn_mtime_t trecv, uint32_t statusinfo)
 {
+#ifndef DDSI_INCLUDE_LIFESPAN
+  DDSRT_UNUSED_ARG (trecv);
+  DDSRT_UNUSED_ARG (statusinfo);
+#endif
   wrinfo->guid = e->guid;
   wrinfo->ownership_strength = xqos->ownership_strength.value;
   wrinfo->auto_dispose = xqos->writer_data_lifecycle.autodispose_unregistered_instances;
   wrinfo->iid = e->iid;
+#ifdef DDSI_INCLUDE_LIFESPAN
+  if ((statusinfo & (NN_STATUSINFO_UNREGISTER | NN_STATUSINFO_DISPOSE)) == 0 && xqos->lifespan.duration != DDS_INFINITY)
```

Comment:
I'd flip the two conjuncts: the primary consideration is whether lifespan is enabled, and only if it is is it of interest whether the thing is a dispose or unregister. (It should be marginally faster on average as well: the vast majority of the stuff is regular data with a lifespan of ∞, so flipping them means you know after one test, rather than after two. That never hurts!)